### PR TITLE
Issue #9021: fix josm validation

### PR DESCRIPTION
**Pull request by pbludov**
Issue #9021

`validation.sh` installs `SNAPSHOT` version of the Checkstyle to the maven cache. But JOSM uses ANT, so it can't find the right jar file. To solve this problem, the `ivysettings.xml` file is patched to use the Checkstyle bundle from the `target` directory.

Before patching:
```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- License: GPL. For details, see LICENSE file. -->
<ivysettings>
  <settings defaultResolver="josm-nexus"/>
  <resolvers>
    <ibiblio name="josm-nexus" m2compatible="true" root="https://josm.openstreetmap.de/nexus/content/repositories/public/"/>
    <ibiblio name="jcenter" m2compatible="true" root="https://jcenter.bintray.com/"/>
  </resolvers>
  <modules>
    <module organisation="ch.poole" name="OpeningHoursParser" resolver="jcenter"/>
  </modules>
</ivysettings>
```
after patching
```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- License: GPL. For details, see LICENSE file. -->
<ivysettings>
  <settings defaultResolver="josm-nexus"/>
  <resolvers>
    <ibiblio name="josm-nexus" m2compatible="true" root="https://josm.openstreetmap.de/nexus/content/repositories/public/"/>
    <ibiblio name="jcenter" m2compatible="true" root="https://jcenter.bintray.com/"/>
    <filesystem name="local-checkstyle">
      <artifact pattern="${base.dir}/../../target/[artifact]-[revision]-all.[ext]"/>
    </filesystem>
  </resolvers>
  <modules>
    <module organisation="ch.poole" name="OpeningHoursParser" resolver="jcenter"/>
    <module organisation="com.puppycrawl.tools" name="checkstyle" resolver="local-checkstyle"/>
  </modules>
</ivysettings>
```

Here we add a new file system resolver, as well as telling ANT to resolve `com.puppycrawl.tools/checkstyle` artifact with this resolver.

### CODE REVIEW

**Comment by romani** [comment]
@pbludov , josm is red in semaphore

**Comment by pbludov** [comment]
> @pbludov , josm is red in semaphore

Now the error is
```
BUILD FAILED01:18
/home/semaphore/checkstyle/.ci-temp/josm/build.xml:894: java.lang.NoClassDefFoundError: antlr/collections/AST
```
I assume that now ANT can find checkstyle jar, but can't find one of its dependencies. This should be reproducible (with clean caches). I will investigate this issue further.

**Comment by pbludov** [comment]
Now the output is
```xml
BUILD SUCCESSFUL
Total time: 35 seconds
Checkstyle Errors:
<error line="543" column="17" severity="warning" message="Array brackets at illegal position." source="com.puppycrawl.tools.checkstyle.checks.ArrayTypeStyleCheck"/>
<error line="544" column="17" severity="warning" message="Array brackets at illegal position." source="com.puppycrawl.tools.checkstyle.checks.ArrayTypeStyleCheck"/>
<error line="546" column="39" severity="warning" message="&apos;,&apos; is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck"/>
<error line="581" column="29" severity="warning" message="&apos;if&apos; is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAroundCheck"/>
<error line="593" column="63" severity="warning" message="&apos;typecast&apos; is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAfterCheck"/>
<error line="594" column="25" severity="warning" message="&apos;if&apos; is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAroundCheck"/>
<error line="597" column="37" severity="warning" message="&apos;if&apos; is not followed by whitespace." source="com.puppycrawl.tools.checkstyle.checks.whitespace.WhitespaceAroundCheck"/>
Size of output: 7
```

**Comment by pbludov** [comment]
All these violations have already been fixed by this [commit](https://josm.openstreetmap.de/changeset/17332/josm)
We just need to wait until the tested version of josm is updated. Right now it is [17329](https://josm.openstreetmap.de/wiki/TestedVersion)

**Comment by romani** [comment]
@pbludov, can we checkout to certain revision ? We really do not care in what revision code we just need to have 0 violation.

**Comment by pbludov** [comment]
> You can comment out checkout to released version and just use hack revision with comment.

done

File: .ci/validation.sh [hunk 1]
```diff
@@ -61,7 +74,7 @@ check-missing-pitests)
 
   echo "List of missing files in pitest profiles: $results"
 
-  if [[ -n $results ]] ; then
+  if [[ -n $results ]]; then
```
Comment (romani) [inline .ci/validation.sh:77]: please move ALL unrelated formatting to separate commit.

what tool do you use for this ?

File: .ci/validation.sh [hunk 2]
```diff
@@ -28,23 +42,22 @@ check-missing-pitests)
   fail=0
   mkdir -p target
 
-  list=($(cat pom.xml | \
+  list=($(cat pom.xml |
```
Comment (romani) [inline .ci/validation.sh:45]: I surprised that `|` at the end imply `\`

File: .ci/validation.sh [hunk 3]
```diff
@@ -76,17 +89,20 @@ eclipse-static-analysis)
...
+    sleep 5s
+    false
+  fi
   rm .ci-temp/output.txt
   ;;
 
 no-error-pmd)
   CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
-                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+    --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
```
Comment (romani) [inline .ci/validation.sh:105]: it looks better in previous version.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .ci/validation.sh (hunk 1 above)
Comment (pbludov) [reply .ci/validation.sh:77]: None. Just Ctrl+X/Ctrl+V in IntellyJ IDEA.

**Comment by pbludov** [comment]
@romani all unexpected changes are reverted.

File: .ci/validation.sh (hunk 1 above)
Comment (pbludov) [reply .ci/validation.sh]: I recently update IDEA to 2020.3, this may be related.
